### PR TITLE
Cosmos-0.42 | getPools & getLpList Pagination

### PR DESCRIPTION
**Pull request by iverc**
- Paginate get pools and get LP list queries

### CODE REVIEW

File: version [hunk 1]
```diff
@@ -1 +1 @@
-0.9.0
+0.9.2
```
Comment (timlind) [inline version:1]: 0.9.1

File: app/setup_handlers.go [hunk 2]
```diff
@@ -7,4 +7,11 @@ import (
 
 func SetupHandlers(app *SifchainApp) {
 	app.UpgradeKeeper.SetUpgradeHandler("0.9.0", func(ctx sdk.Context, plan types.Plan) {})
+	app.UpgradeKeeper.SetUpgradeHandler("0.9.1", func(ctx sdk.Context, plan types.Plan) {})
+	app.UpgradeKeeper.SetUpgradeHandler("0.9.2", func(ctx sdk.Context, plan types.Plan) {})
```
Comment (timlind) [inline app/setup_handlers.go:11]: remove all except 0.9.1

File: x/clp/keeper/liquidityprovider.go [hunk 3]
```diff
@@ -71,21 +78,66 @@ func (k Keeper) DestroyLiquidityProvider(ctx sdk.Context, symbol string, lpAddre
 	store.Delete(key)
 }
 
-func (k Keeper) GetLiquidityProvidersForAsset(ctx sdk.Context, asset types.Asset) []types.LiquidityProvider {
-	var lpList []types.LiquidityProvider
+func (k Keeper) GetLiquidityProvidersForAsset(ctx sdk.Context, asset types.Asset) []*types.LiquidityProvider {
```
Comment (timlind) [inline x/clp/keeper/liquidityprovider.go:81]: Is this still used?

File: x/clp/keeper/liquidityprovider.go [hunk 4]
```diff
@@ -71,21 +78,66 @@ func (k Keeper) DestroyLiquidityProvider(ctx sdk.Context, symbol string, lpAddre
...
+func (k Keeper) GetLiquidityProvidersForAssetPaginated(ctx sdk.Context, asset types.Asset, pagination *query.PageRequest) ([]*types.LiquidityProvider, *query.PageResponse, error) {
+	var lpList []*types.LiquidityProvider
+	logger := k.Logger(ctx).With("func", "GetLiquidityProvidersForAssetPaginated")
+
+	store := ctx.KVStore(k.storeKey)
+	lpStore := prefix.NewStore(store, types.LiquidityProviderPrefix)
+
+	pageRes, err := query.FilteredPaginate(lpStore, pagination, func(key []byte, value []byte, accumulate bool) (bool, error) {
+		var lp types.LiquidityProvider
+		logger.Info(fmt.Sprintf("attempting to load liquidity provider for %s", key))
```
Comment (timlind) [inline x/clp/keeper/liquidityprovider.go:105]: can remove this log.

File: x/clp/keeper/liquidityprovider.go [hunk 5]
```diff
@@ -71,21 +78,66 @@ func (k Keeper) DestroyLiquidityProvider(ctx sdk.Context, symbol string, lpAddre
...
+			lpList = append(lpList, &lp)
+		}
+
+		return true, nil
+	})
+	if err != nil {
+		return nil, &query.PageResponse{}, status.Error(codes.Internal, err.Error())
+	}
+
+	logger.Info(fmt.Sprintf("returning liquidity provider for asset %s with %d results", asset.Symbol, pageRes.Total))
```
Comment (timlind) [inline x/clp/keeper/liquidityprovider.go:136]: Can remove this log.

File: x/clp/keeper/pool.go [hunk 6]
```diff
@@ -64,6 +65,26 @@ func (k Keeper) GetPools(ctx sdk.Context) []*types.Pool {
 	return poolList
 }
 
+func (k Keeper) GetPoolsPaginated(ctx sdk.Context, pagination *query.PageRequest) ([]*types.Pool, *query.PageResponse, error) {
```
Comment (timlind) [inline x/clp/keeper/pool.go:68]: Is there a function this deprecates that can be removed?

File: x/clp/keeper/querier.go [hunk 7]
```diff
@@ -23,35 +22,29 @@ func NewQuerier(keeper Keeper, legacyQuerierCdc *codec.LegacyAmino) sdk.Querier
 		case types.QueryAssetList:
 			return queryAssetList(ctx, path[1:], req, keeper, legacyQuerierCdc)
 		case types.QueryLPList:
-			return queryLPList(ctx, path[1:], req, keeper, legacyQuerierCdc)
+			return queryLPList(ctx, path[1:], req, keeper, legacyQuerierCdc, querier)
 		case types.QueryAllLP:
-			return queryAllLP(ctx, path[1:], keeper, legacyQuerierCdc)
+			return queryAllLP(ctx, path[1:], keeper, legacyQuerierCdc, querier)
```
Comment (timlind) [inline x/clp/keeper/querier.go:27]: Get pools should also be hooked up to go through query server.

**Review by timlind — CHANGES_REQUESTED** [review]
I still need to take a look at the changes not related to the fix.

File: x/clp/keeper/liquidityprovider.go [hunk 8]
```diff
@@ -71,21 +73,53 @@ func (k Keeper) DestroyLiquidityProvider(ctx sdk.Context, symbol string, lpAddre
 	store.Delete(key)
 }
 
-func (k Keeper) GetLiquidityProvidersForAsset(ctx sdk.Context, asset types.Asset) []types.LiquidityProvider {
-	var lpList []types.LiquidityProvider
+// Deprecated: GetLiquidityProvidersForAsset use GetLiquidityProvidersForAssetPaginated
+func (k Keeper) GetLiquidityProvidersForAsset(ctx sdk.Context, asset types.Asset) []*types.LiquidityProvider {
+	var lpList []*types.LiquidityProvider
```
[the remark this answers is not in the document]
Comment (iverc) [reply x/clp/keeper/liquidityprovider.go:78]: @timlind  yeah, should I remove it or "Deprecated" mark is ok?

File: x/clp/keeper/querier.go (hunk 7 above)
Comment (iverc) [reply x/clp/keeper/querier.go:27]: they are

File: x/clp/keeper/grpc_query.go [hunk 9]
```diff
@@ -105,54 +92,30 @@ func (k Querier) GetLiquidityProviderList(c context.Context, req *types.Liquidit
...
-		}
-
-		if accumulate {
-			lpl = append(lpl, &lp)
-		}
-
-		return true, nil
-	})
-
+	lpList, pageRes, err := k.Keeper.GetAllLiquidityProvidersPaginated(ctx, req.Pagination)
```
[the remark this answers is not in the document]
Comment (iverc) [reply x/clp/keeper/grpc_query.go:113]: if not provided, default limit (100) is used

File: x/clp/keeper/grpc_query.go [hunk 10]
```diff
@@ -105,54 +92,30 @@ func (k Querier) GetLiquidityProviderList(c context.Context, req *types.Liquidit
...
 	ctx := sdk.UnwrapSDKContext(c)
-
 	searchingAsset := types.NewAsset(req.Symbol)
-	lpList := k.GetLiquidityProvidersForAsset(ctx, searchingAsset)
-
-	lpl := make([]*types.LiquidityProvider, len(lpList))
-	for i, lp := range lpList {
-		lp := lp
-		lpl[i] = &lp
+	lpList, pageRes, err := k.Keeper.GetLiquidityProvidersForAssetPaginated(ctx, searchingAsset, req.Pagination)
```
[the remark this answers is not in the document]
Comment (iverc) [reply x/clp/keeper/grpc_query.go:97]: if not provided, default limit (100) is used

File: x/clp/keeper/grpc_query.go [hunk 11]
```diff
@@ -42,25 +37,24 @@ func (k Querier) GetPools(c context.Context, req *types.PoolsReq) (*types.PoolsR
 	if req == nil {
 		return nil, status.Error(codes.InvalidArgument, "empty request")
 	}
-
 	ctx := sdk.UnwrapSDKContext(c)
-
-	pools := k.Keeper.GetPools(ctx)
-
+	pools, pageRes, err := k.Keeper.GetPoolsPaginated(ctx, req.Pagination)
```
[the remark this answers is not in the document]
Comment (iverc) [reply x/clp/keeper/grpc_query.go:41]: if not provided, default limit (100) is used

File: x/clp/genesis.go [hunk 12]
```diff
@@ -41,24 +38,19 @@ func InitGenesis(ctx sdk.Context, k keeper.Keeper, data types.GenesisState) (res
...
 func ExportGenesis(ctx sdk.Context, keeper keeper.Keeper) types.GenesisState {
 	params := keeper.GetParams(ctx)
-
-	poolList := keeper.GetPools(ctx)
-
-	liquidityProviders := keeper.GetLiquidityProviders(ctx)
-
+	var poolList []*types.Pool
+	poolList, _, _ = keeper.GetPoolsPaginated(ctx, &query.PageRequest{})
+	liquidityProviders, _, _ := keeper.GetAllLiquidityProvidersPaginated(ctx, &query.PageRequest{})
```
[the remark this answers is not in the document]
Comment (iverc) [reply x/clp/genesis.go:48]: we can set limit to max uint here to get all records

File: x/clp/keeper/querier_test.go [hunk 13]
```diff
@@ -224,36 +208,30 @@ func TestQueryAssetList(t *testing.T) {
...
-
-	resBz, err := querier(ctx, []string{types.QueryLPList}, query)
-	require.NoError(t, err)
-
+	qLpList, err := querier(ctx, []string{types.QueryLPList}, query)
+	assert.NoError(t, err)
 	res := types.LiquidityProviderListRes{}
-	err = cdc.UnmarshalJSON(resBz, &res.LiquidityProviders)
-	require.NoError(t, err)
+	err = cdc.UnmarshalJSON(qLpList, &res)
```
Comment (timlind) [inline x/clp/keeper/querier_test.go:228]: This changes the JSON format returned to caller, breaking clients.

File: x/clp/keeper/querier_test.go [hunk 14]
```diff
@@ -262,20 +240,14 @@ func TestQueryAllLPs(t *testing.T) {
...
 	resBz, err := querier(ctx, []string{types.QueryAllLP}, query)
-	require.NoError(t, err)
-
-	res := types.LiquidityProvidersRes{}
-	err = cdc.UnmarshalJSON(resBz, &res.LiquidityProviders)
-	require.NoError(t, err)
-	require.Equal(t, []*types.LiquidityProvider{&lp}, res.LiquidityProviders)
+	assert.NoError(t, err)
+	var lpRes types.LiquidityProvidersRes
+	err = cdc.UnmarshalJSON(resBz, &lpRes)
```
Comment (timlind) [inline x/clp/keeper/querier_test.go:248]: Also a breaking change.

**Review by timlind — DISMISSED** [review]

**Pull request merged**